### PR TITLE
NXdata errors on an axis, signal or auxiliary signal

### LINKED ISSUE #1044: NXdata: errors on auxiliary signals
We have the `VARIABLE_errors` fields to specify the errors on all axes but we only have one `errors` field to specify the errors on the signal. I would propose `DATA_errors` to specify errors on all signals (auxiliary and default).

```
data:NXdata
   @axes=["x"]
   @signal="data1"
   @auxiliary_signals=["data2", "data3"]
   x: [1024]
   data1: [1024]
   data2: [1024]
   data3: [1024]
   x_errors: [1024]
   data1_errors: [1024]
   data2_errors: [1024]
   data3_errors: [1024]
```

`data1_errors` is `errors` in the current definition and `data2_errors`, `data3_errors` do not exist in the current definition.

### PR DESCRIPTION
**Pull request by woutdenolf**
Closes #1044 

Currently [`NXdata`](https://manual.nexusformat.org/classes/base_classes/NXdata.html) does not allow errors on auxiliary signals.

This PR solves that by
* renaming the `VARIABLE` field to `AXISNAME`
* updating the doc of `VARIABLE_errors` to say it applies to axes, signal or auxiliary signals: this is the actual change, although it is already somewhat support by NeXus as described [here](https://manual.nexusformat.org/datarules.html?highlight=errors).
* deprecate `errors` in favor of `VARIABLE_errors`  for the signal field

### CODE REVIEW

**Comment by woutdenolf** [comment]
For example

```
test.h5:NXroot
  @NX_class = "NXroot"
  @default = "entry"
  data:NXdata
    @NX_class = "NXdata"
    @auxiliary_signals = ["I1", "I2"]
    @axes = ["energy"]
    @signal = "I0"
    I0:int64[3] = [1, 0, 1]
    I0_errors:float64[3] = [0.1, 0.0, 0.1]  # called "errors" in current definition
    I1:int64[3] = [1, 0, 1]
    I1_errors:float64[3] = [0.1, 0.0, 0.1] # NOT in current definition
    I2:int64[3] = [1, 0, 1]
    I2_errors:float64[3] = [0.1, 0.0, 0.1] # NOT in current definition
    energy:int64[3] = [1, 5, 2]
    energy_errors:float64[3] = [0.1, 0.4, 0.15]
  entry:NXroot
    @NX_class = "NXroot"
    @default = "data"
```


```python
import h5py

with h5py.File("test.h5", "w") as f:
    f.attrs["NX_class"] = "NXroot"
    f.attrs["default"] = "entry"

    entry = f.create_group("entry")
    entry.attrs["NX_class"] = "NXroot"
    entry.attrs["default"] = "data"

    data = f.create_group("data")
    data.attrs["NX_class"] = "NXdata"
    data.attrs["axes"] = ["energy"]
    data.attrs["signal"] = "I0"
    data.attrs["auxiliary_signals"] = ["I1", "I2"]

    data["energy"] = [1, 5, 2]
    data["energy_errors"] = [0.1, 0.4, 0.15]

    data["I0"] = [1, 0, 1]
    data["I0_errors"] = [0.1, 0, 0.1]  # called "errors" in current definition

    data["I1"] = [1, 0, 1]
    data["I1_errors"] = [0.1, 0, 0.1]  # NOT in current definition

    data["I2"] = [1, 0, 1]
    data["I2_errors"] = [0.1, 0, 0.1]  # NOT in current definition
```

**Comment by vasole** [comment]
Just to add on it, that is something already found in the documentation at:

 https://manual.nexusformat.org/datarules.html?highlight=_errors

but not found when describing NXdata. Here is the information concerning reserved suffixes


<html><body>
<p class="rubric" id="reserved-suffixes">Reserved suffixes</p>
<p>When naming a field, NeXus has reserved certain suffixes to the names
so that a specific meaning may be attached.  Consider a field named <code class="docutils literal notranslate"><span class="pre">DATASET</span></code>,
the following table lists the suffixes reserved by NeXus.</p>


suffix | reference | meaning
-- | -- | --
_end | NXtransformations | end points of the motions that start with DATASET
_errors | NXdata | uncertainties (a.k.a., errors)
_increment_set | NXtransformations | intended average range through which the corresponding axis moves during the exposure of a frame
_indices | NXdata | Integer array that defines the indices of the signal field which need to be used in the DATASET in order to reference the corresponding axis value
_mask |   | Field containing a signal mask, where 0 means the pixel is not masked. If required, bit masks are defined in NXdetector pixel_mask.
_set | target values | Target value of DATASET
_weights |   | divide DATASET by these weights [4]


<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id6" role="note">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="https://manual.nexusformat.org/datarules.html?highlight=_errors#id5">4</a><span class="fn-bracket">]</span></span>
<p>If <code class="docutils literal notranslate"><span class="pre">DATASET_weights</span></code> exists and has the same shape as the field,
you are supposed to divide <code class="docutils literal notranslate"><span class="pre">DATASET</span></code> by the weights.</p>
</aside>
</aside>
</body>
</html>

**Comment by benajamin** [comment]
We seem to have already provided 2 different general strategies for giving errors, as described in https://www.nexusformat.org/NIACDesign.html

Maybe we need to decide between `_errors` and `@uncertainties`

**Comment by prjemian** [comment]
As I recall, this was decide at a previous NIAC to use `_errors` and deprecate use of `@uncertainties`.  The documentation may be lagging and should be edited to current decision.

**Comment by woutdenolf** [comment]
I propose to leave `_errors` vs `@uncertainties` for another discussion. For example here: https://github.com/nexusformat/definitions/issues/1185

This MR deals specifically with make it clear in NXdata how the add errors to signal, axes and auxiliary signals.

**Comment by prjemian** [comment]
Suggestion to deprecate the `errors` field since it is superseded by the `_errors` suffix.  (Undefined behavior of both present.)

**Comment by mkoennecke** [comment]
The link to the NIAC 2018 minutes: https://www.nexusformat.org/NIAC2018Minutes.html

**Review by prjemian — APPROVED** [review]
Looks right to me.  Consistent with NIAC2018.

File: base_classes/NXdata.nxdl.xml [hunk 1]
```diff
@@ -306,24 +306,13 @@
...
-			throughout a NeXus data file to associate uncertainties 
-			with a field named ``VARIABLE``.  This pattern also 
-			applies to other relationships such as ``VARIABLE_resolutions``
-			to connect additional data with a certain field. 
+			"Errors" (meaning *uncertainties* or *standard deviations*)
+			associated with any field named ``VARIABLE`` in this ``NXdata``
+			group (e.g. an axis, signal or auxiliary signal).
+
+			The dimensions of the ``VARIABLE_errors`` field must match
+			the dimensions of the ``VARIABLE`` field.
```
Comment (woutdenolf) [inline base_classes/NXdata.nxdl.xml:314]: Perhaps this?

```
"Errors" (meaning *uncertainties* or *standard deviations*)
associated with any field named ``VARIABLE`` (``AXISNAME`` or ``DATA``) in this ``NXdata``
group (e.g. an axis, signal or auxiliary signal).

The dimensions of the ``VARIABLE_errors`` field must match
the dimensions of the ``VARIABLE`` field.
```

Comment (woutdenolf) [reply base_classes/NXdata.nxdl.xml:314]: Or perhaps two fields?

```
<field name="AXISNAME_errors" type="NX_NUMBER" nameType="any">
<field name="DATA_errors" type="NX_NUMBER" nameType="any">
```

File: base_classes/NXdata.nxdl.xml [hunk 2]
```diff
@@ -306,24 +306,13 @@
 	</field>
 	<field name="VARIABLE_errors" type="NX_NUMBER" nameType="any">
```
Comment (benajamin) [inline base_classes/NXdata.nxdl.xml:307]: Replacing "VARIABLE" with "FIELDNAME" would be more general and easier to understand.

File: base_classes/NXdata.nxdl.xml (hunk 1 above)
Comment (woutdenolf) [reply base_classes/NXdata.nxdl.xml]: Using ``FIELDNAME`` instead of ``VARIABLE``

**Review by PeterC-DLS — APPROVED** [review]
Happy with this

**Comment by benajamin** [comment]
Just to clarify the discussion here, there was a decision at the 2018 NIAC meeting:

> The errors field and uncertainties attribute in NXdata to be marked deprecated

Therefore issue #1044 is moot and only a clarification of the documentation is required, which was accomplished by this pull request.

**Pull request merged**